test/SmrTest/lib/DefaultGame/MySqlDatabaseTest.php

```diff
@@ -0,0 +1,47 @@
+<?php
+
+namespace SmrTest\lib\DefaultGame;
+
+use DI\Container;
+use MySqlDatabase;
+use mysqli;
+use PHPUnit\Framework\MockObject\MockObject;
+use PHPUnit\Framework\TestCase;
+use RuntimeException;
+use Smr\Container\DiContainer;
+
+class MySqlDatabaseTest extends TestCase {
+	private Container $container;
+	/**
+	 * @var mysqli|MockObject
+	 */
+	private mysqli $mysql;
+
+	protected function setUp(): void {
+		DiContainer::initializeContainer();
+		$this->container = DiContainer::getContainer();
+		$this->mysql = $this->createMock(mysqli::class);
+		// Replace the factory definition for mysqli object with our mock, so when
+		// requesting a mysqli instance, it will always return a mock for this test
+		$this->container->set(mysqli::class, $this->mysql);
+	}
+
+	public function test__construct_happy_path() {
+		$this->mysql
+			->expects(self::once())
```

Comment:
I'm working on a way to fix Scrutinizer on this, but I'm not sure how just yet. Unfortunately PHP7 doesn't seem to support union types. I tried fixing it with `@var` in the PHPDoc, but that didn't satisfy The Scrute.

I may end up switching up the testing infrastructure in favour of [PHPSpock's PHPUnit adapter](https://github.com/aleczhang/PhpSpock). Then I can write these tests like I write [Spock tests](http://spockframework.org/spock/docs/1.3/all_in_one.html) in the JVM world. Spock tests are really slick, and easy to read without near as much setup.

Reply:
PHP8 will have union types, and it is released in just over a week (Nov 26, 2020). I will be upgrading the containers to PHP8 as soon as I can. Will that solve this problem?

Looking at the PhpSpock library, it's barely been touched since it was created 9 years ago (and your PR is the only PR ever!). I'm not sure I would be comfortable adding a dead dependency.

Reply:
Union types should solve the problem, I think.

Another issue is that the base PHPUnit mocking/stubbing does not lent itself well to the refactoring of method names. In the expectation chain you declare a method to be called. If this method gets renamed, or moved, the tests to not reflect the change because you are using a string literal.

With PhpSpock, the method / fields that are stubbed out / verified are always going to be correct, as the object that gets returned is a correctly typehinted object (not some mock implementation). I don't know if you've ever had the pleasure of working with Spock for unit tests in the past, but it's the cleanest, most efficient testing framework I've ever used.

If something works, and works well it doesn't really need to be updated. As long as it fulfils its purpose, I have no problem with using old dependencies (especially a test dependency). The only issue that I've experienced with PhpSpock is that the PHPUnit namespaces are old, but I resolved them locally and created several tests using their adapter, and it worked gloriously.

Reply:
Are you vehemently opposed to me using PhpSpock? I will use my fork so it's updated

Reply:
The benefits of Spock sound nice. I feel similarly about `unittest` vs. `pytest` when I'm working in python. I am still worried about the fact that the project is abandoned, because everything abandoned breaks at some point. I guess the big questions are: a) who is going to migrate back to PHPUnit when it breaks and b) how much effort will it be to do that? If the answers are "@hemberger" and "a lot", then I would definitely want to stick with PHPUnit. What do you think?

Reply:
It's still using PHPUnit

Reply:
I saw that PhpSpock had PHPUnit as a dependency. Can you show me an example of what the tests would look like with & without PhpSpock?

Reply:
It doesn't have a dependency on PHPUnit -- you have to provide that yourself if you want to use the adapter:

Example (base PHP):
```php
public function test_validate_config_happy_path() {
	//# Given a Dotenv object
	$dotEnv = $this->createMock(Dotenv::class);
	// And the dotenv config will return the following array when loaded
	$dotEnv
		->expects(self::once())
		->method("load")
		->willReturn([
			"MYSQL_HOST" => "host",
			"MYSQL_USER" => "user",
			"MYSQL_PASSWORD" => "pass",
			"MYSQL_DATABASE" => "database"
		]);
	// And we expect that the dotenv "required" method will be called with the following arguments
	$dotEnv
		->expects(self::exactly(4))
		->method("required")
		->with(
			$this->logicalOr(
				"MYSQL_USER",
				"MYSQL_HOST",
				"MYSQL_PASSWORD",
				"MYSQL_DATABASE"));
	// When constructing the properties class
	$mysqlProperties = new MySqlProperties($dotEnv);
	// Then the properties have expected values
	self::assertEquals("host", $mysqlProperties->getHost());
	self::assertEquals("user", $mysqlProperties->getUser());
	self::assertEquals("pass", $mysqlProperties->getPassword());
	self::assertEquals("database", $mysqlProperties->getDatabaseName());
}
```

Example Spock:
```php
public function test_validate_config_happy_path_using_spock() {
	/**
	 * @var $dotEnv \Dotenv\Dotenv *Mock*
	 */
	setup:
	1 * $dotEnv->load() >> array(
		"MYSQL_HOST" => "host",
		"MYSQL_USER" => "user",
		"MYSQL_PASSWORD" => "pass",
		"MYSQL_DATABASE" => "database");
	1 * $dotEnv->required("MYSQL_USER");
	1 * $dotEnv->required("MYSQL_HOST");
	1 * $dotEnv->required("MYSQL_PASSWORD");
	1 * $dotEnv->required("MYSQL_DATABASE");
	when:
	$mysqlProperties = new MySqlProperties($dotEnv);
	then:
	$mysqlProperties->getHost() == "host";
	$mysqlProperties->getUser() == "user";
	$mysqlProperties->getPassword() == "pass";
	$mysqlProperties->getDatabaseName() == "database";
}
```